php/php-k8s-v8/etc/nginx/apps/drupal/drupal.conf

```diff
@@ -296,15 +310,52 @@ location @drupal-no-args {
     #proxy_set_header Connection '';
 }
 
-## Internal location to ask Drupal to generate a derivative image if the source
-## image is present.
+## Internal location to ask Drupal to generate a derivative image if the
+## source ## image is present. Add a bit of logic so that it does not break
+## stage_file_proxy on non-prod environments.
 location @drupal-generate-derivative-image {
-  ## If the source image doesn't exist return a 404.
+
+  set $allow_generate "";
+  set_by_lua $environment 'return os.getenv("ENVIRONMENT")';
+
+  # Set a flag if this is the production environment, so we can handle an
+  # exception for stage_file_proxy.
+  if ($environment = "prod") {
+    set $allow_generate "P";
+  }
+
+  # Set a flag if the source file does not exist.
   if (!-f "$document_root/sites/default/files/$file_path") {
+    set $allow_generate "${allow_generate}F";
+  }
+
+  # Set a flag if the request is *not* for a webp image.
+  if ($file_path ~* "^.*(?<!webp)$") {
+    set $allow_generate "${allow_generate}W";
+  }
+
+  # If production *and* the source file does not exist, return a hard nope
+  # for this derivative image request.
+  if ($allow_generate = "PFW") {
+    return 404;
+  }
+
+  # Otherwise pass the request to drupal so that stage_file_proxy and
+  # image styles can do their thing.
+  try_files /dev/null @drupal;
```

Comment:
```suggestion
  set_by_lua $environment 'return os.getenv("ENVIRONMENT")';

  # If this is not the production environement, pass the request to Drupal.
  if ($environment != "prod") {
    return 406;
  }

  # If the image is a webp image, pass the request to Drupal.
  if ($file_path ~* "^.*\.webp$") {
    return 406;
  }

  # If the source image doesn't exist, return a 404.
  if (!-f "$document_root/sites/default/files/$file_path") {
    return 404;
  }

  # Otherwise, pass the request to Drupal to generate the derivative.
  try_files /dev/null @drupal;
```

I couldn't get the correct behavior and was always getting a nginx 404 with the flag system because nginx seems to change the context when a "if" matches. So when the "allow_generate" was not a "PFW", it returned a 404 without passing the request to drupal.

The above suggestion, that seems to work according to my tests, reverses the logic and always passes the request to drupal if the environment is not prod or the image is a webp or the source image exists.

According to https://www.nginx.com/resources/wiki/start/topics/depth/ifisevil/, only "return" and "rewrite... last" as 100% safe inside "if" in a location context.

Reply:
Oh, yes, nice catch. I'd forgotten about that 406 status handler :-)